### DISCUSSION TITLE
Getting Reccuring transactions setup correctly

### DISCUSSION THREAD

**Discussion by geomule** [firefly-iii/firefly-iii#9582] [Firefly III] [answered]
Hi, I've achieved a big milestone by installing FF and really like it. Now my challenge is getting reoccurring transaction (this post) and import (a separate post) going. 

Setup:
* Debian 12 VM running on server.
* FF 6.1.25 running as container on the above VM which is attached to 192.168.100.5:84
* Set up token under APP_KEY om the .env file
*  I have the volume set as following in the yml file:
    volumes:
      - firefly_iii_upload:/var/www/html/storage/upload

I see the alpine cron container running in Portainer with healthy status.
As far as I understand, all I need is to set up the cron job as per [here](https://docs.firefly-iii.org/how-to/firefly-iii/advanced/cron/). 

I created the cronjob with `sudo crontab -u www-data -e` and added the following in the editor:
    
    # cron job for Firefly III
    0 3 * * * /usr/bin/php /var/www/html/artisan firefly-iii:cron

What I'm struggling to understand is if this is actually working. What I understand is that by calling `sudo crontab -u www-data -e`, I'm doing this through the bash terminal of the VM and therefore the `/var/www/html/artisan` on the VM filesystem. But when I cd to the `/var/www/html/`, there is nothing in there. Presumably this is because the path refers to the container' volume?

When cron daily is executed, php of VM won't find any `artisan` in the `/var/www/html/`, so how can it execute the cron?
I'm no expert of docker so clearly I'm missing something here. Any helpful tip will certainly help. Is there a way of testing these reoccuring transaction without having to wait for the rule to be true? Is it with the `--force --date=$(date "+\%Y-\%m-")10` and then checking the logs of the cron container in Portainer?

Thanks

**Comment by JC5** [comment]
Mmmm, that cron job will not work. Like you say, there is nothing in that path.

What you need is a little further down the page: https://docs.firefly-iii.org/how-to/firefly-iii/advanced/cron/#from-outside-the-container-cli

My own cron job command is, for example:

```
0 1 * * * docker exec firefly-iii-app-1 php artisan firefly-iii:cron
```

It is slightly different from the manual, because I know what the container will be named.

**Reply by geomule** [reply]
Ok, let me se if I can unpack this as my Bash skills a limited. The webpage you reference says to run this.

```
# cron job for Firefly III using Docker itself
  0 3 * * * docker exec $(docker container ls -a -f name=firefly --format="{{.ID}}") /usr/local/bin/php /var/www/html/artisan firefly-iii:cron
```

Now I need to understand what those bits are so I can personalise to my situation. Note that I have not change the name of the containers so have used you docker compose file as is, just changing the port to avoid conflicts with other containers.

1. This bit   `0 3 * * * docker exec` is just telling docker to execute what is in brackets.
2. This bit in brackets `$(docker container ls -a -f name=firefly --format="{{.ID}}")` is telling `docker exec` to find all the containers that have the name `firefly` and to determine the ID of that container? But I have 4 containers with that name in there (see image below). In fact, if I run that command in the brackets in the bash shell, i get 4 IDs returned. Your command doesn't have this bit in brackets. Why?
3. This bit `/usr/local/bin/php /var/www/html/artisan firefly-iii:cron` is the bit that `docker exec` actually runs; basically the PHP binary in the container is to run the `artisan` which in turn is having to find the rules somewhere with the `firefly-iii:cron`? Is this last bit the container name?

![immagine](https://github.com/user-attachments/assets/56718b89-e6ec-47b1-bc9b-11ea920ac44c)

**Reply by JC5** [reply]
I'm pretty sure you can solve all of these riddles yourself.

**Reply by geomule** [reply]
I appreciate you taking the time to read my posts. While I understand that my questions may seem trivial,  they are less so for those like me that understand little about intricacies of docker environment. I do try experimenting, trust me :-)

This is how far I've got without success. After launching `docker ps -a`, I get the IDs below

```
f30b3156467e   fireflyiii/data-importer:latest
9a3895b857fe   fireflyiii/core:latest                  
b0a37ab9dc2e   alpine 
e07c7172f168   mariadb:lts 
```
Then I launch this from Bash of VM `docker exec b0a37ab9dc2e /usr/local/bin/php artisan 9a3895b857fe:cron --force --date=2025-01-26` but it complains now that it can't find the PHP:

```OCI runtime exec failed: exec failed: unable to start container process: exec: "/usr/local/bin/php": stat /usr/local/bin/php: no such file or directory: unknown```

As you can see from the above, I assumed that the `firefly_iii_cron` has to launch `firefly_iii_core`. I've given up using container names as no matter what I try, it complains the container does not exist. With IDs, seems to work.

**Reply by JC5** [reply]
OK so when you run a command in `$()`, it returns the output. That's useful to get a small command output as a variable. But you don't need all of that.


Since you're running the cron container, something I've just realized (sorry), it should be enough to set an environment variable called `STATIC_CRON_TOKEN` and give it the same value as it has in the docker configuration. For example, use `STATIC_CRON_TOKEN=Ft7BghmqRhJpK6Dk7YDincBPvfy05ab5`.

Then in the Docker compose file:

https://github.com/firefly-iii/docker/blob/main/docker-compose.yml#L38

REPLACEME must get the same value.

For example:

```
      | echo \"0 3 * * * wget -qO- http://app:8080/api/v1/cron/Ft7BghmqRhJpK6Dk7YDincBPvfy05ab5;echo\" 
```

If you do this, the cron container will pick up the cron job for you.

**Reply by geomule** [reply, marked as answer]
Ok, thanks JC5. I think I got this working although I need to see if new transactions are created tomorrow as I still get errors when I try to run manually. :-) But now at least I understand that the cron container does all the work for you and you don't need to set anything other that the TOKEN in the .env and the docker compose files.

Using the image under "Docker compose" [here](https://docs.firefly-iii.org/how-to/firefly-iii/advanced/cron/#from-outside-the-container-cli), I modified the command a bit to `docker container ls --format '{{.ID}}\t{{.Image}}:\t{{.Command}}' | grep alpine` which gives me:

```
name@machine:~$ docker container ls --format '{{.ID}}\t{{.Image}}:\t{{.Command}}' | grep alpine
b0a37ab9dc2e    alpine: "sh -c 'echo \"0 3 * …"
```
and then run:
```

name@machine:~$ docker logs b0a37ab9dc2e
crond: crond (busybox 1.37.0) started, log level 8
crond: USER root pid   9 cmd wget -qO- http://app:8080/api/v1/cron/MY32CHARTOKENHERE;echo
wget: server returned error: HTTP/1.1 404 Not Found

crond: USER root pid  11 cmd wget -qO- http://app:8080/api/v1/cron/MY32CHARTOKENHERE;echo
wget: server returned error: HTTP/1.1 404 Not Found

crond: USER root pid  13 cmd wget -qO- http://app:8080/api/v1/cron/MY32CHARTOKENHERE;echo
wget: server returned error: HTTP/1.1 404 Not Found

crond: USER root pid  15 cmd wget -qO- http://app:8080/api/v1/cron/MY32CHARTOKENHERE;echo
wget: server returned error: HTTP/1.1 404 Not Found

crond: crond (busybox 1.37.0) started, log level 8
crond: USER root pid  14 cmd wget -qO- http://app:8080/api/v1/cron/MY32CHARTOKENHERE;echo
{"recurring_transactions":{"job_fired":true,"job_succeeded":true,"job_errored":false,"message":"Recurring transactions cron job fired successfully."},"auto_budgets":{"job_fired":true,"job_succeeded":true,"job_errored":false,"message":"Auto-budget cron job fired successfully."},"bill_warnings":{"job_fired":true,"job_succeeded":true,"job_errored":false,"message":"Bill warning cron job fired successfully."}}
```
There is no time/date on the output but I'm assuming the last bit `crond: USER root pid  14 ` is encouraging as is probably the message resulting from the cron run from 3am this morning given I made the change that JC5 advised yesterday (i.e. to add TOKEN in .env and docker compose file.

Hope this thread is useful to others too.  BTW, the option to assign tags all over in Firefly is really great. I've just assigned the INBOX tag to all my Reoccurring Transaction rules so that I can easily filter newly/auto transactions that appears since I last logged in to Firefly.

**Comment by JC5** [comment]
> Is there a way of testing these reoccuring transaction without having to wait for the rule to be true?

Yes, you can force the cron job to run on a certain time or force it manually. For example, you can just type this on the command line and see what happens.

```
docker exec firefly-iii-app-1 php artisan firefly-iii:cron --force --date=2024-12-27
```

Of course, change the date to a date you expect the recurring transaction to be created.
